graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -451,127 +455,215 @@ def remove_backtrack_blocks_from_fold(folded_ir_blocks):
+        # We discard the remainder of the match traversal (everything following is also omitted).
+        if new_step is None:
+            break
+        else:
+            new_match_traversal.append(new_step)
+
+    return new_match_traversal
+
+
+def convert_optional_traversals_to_compound_match_query(
+        match_query, optional_locations, location_to_optional):
     """Return 2^n distinct MatchQuery objects in a CompoundMatchQuery.
 
+    Given a MatchQuery containing `n` optional traverses that expand vertex fields,
+    construct `2^n` different MatchQuery objects:
+    one for each possible subset of optional edges that can be followed.
+    For each edge `e` in a subset of optional edges chosen to be omitted,
+    discard all traversals following `e`, and add filters specifying that `e` *does not exist*.
+
     Args:
-        match_query: MatchQuery object potentially containing n `@optional` scopes
-            which expand vertex fields
+        match_query: MatchQuery object containing n `@optional` scopes which expand vertex fields
+        optional_locations: list of @optional locations (location preceding an @optional traverse)
+            that expand vertex fields within
+        location_to_optional: dict mapping all locations within optional scopes
+            to the corresponding optional location
 
     Returns:
         CompoundMatchQuery object containing 2^n MatchQuery objects,
         one for each possible subset of the n optional edges being followed
     """
-    def filter_local_field_existence(field_name):
-        """Return an Expression that is True iff `field_name` does not exist."""
-        local_field = LocalField(field_name)
-        local_field_size = UnaryTransformation(u'size', local_field)
-        field_null_check = BinaryComposition(u'=', local_field, NullLiteral)
-        field_size_check = BinaryComposition(u'=', local_field_size, ZeroLiteral)
-        return BinaryComposition(u'||', field_null_check, field_size_check)
-
-    def is_optional_traverse(traverse):
-        """Return True if `traverse` contains any traversal within an @optional scope."""
-        return any(
-            isinstance(step.root_block, Traverse) and step.root_block.within_optional_scope
-            for step in traverse
+    optional_location_combinations_list = [
+        itertools.combinations(optional_locations, x)
+        for x in range(0, len(optional_locations) + 1)
+    ]
+    optional_location_subsets = list(itertools.chain(*optional_location_combinations_list))
+
+    compound_match_traversals = []
+    for omitted_locations in reversed(optional_location_subsets):
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            location = match_traversal[0].as_block.location
+            location_in_dict = location in location_to_optional
+
+            if not location_in_dict or location_to_optional[location] not in omitted_locations:
```

Comment:
This seems to be checking that `location` exists within `location_to_optional`, then checking the value grabbed from that dict. Since `location_to_optional` is not allowed to have `None` values, why not write as:

```
optional_location = location_to_optional.get(location, None)
if optional_location is None or optional_location not in omitted_locations:
    ...
``